torch/_inductor/codecache.py

```diff
@@ -249,35 +253,151 @@ def code_hash(code):
+            # Special case for example_inputs, everything in torch/csrc/dynamo/guards.cpp
+            # Except for requires_grad, as that has already been handled
+            pickled_guards = []
+            for t in arg:
+                pickled_guards.append(
+                    b"".join(
+                        [
+                            pickle.dumps(torch._C._dispatch_keys(t).__repr__()),
+                            pickle.dumps(t.dtype),
+                            pickle.dumps(t.device),
+                            pickle.dumps(t.shape),
+                        ]
+                    )
+                )
+            return b"".join(pickled_guards)
+    return pickle.dumps(arg)
+
+
+def compiled_fx_graph_hash(fx_args: List[Any]):
+    serialized_fx_args = b"".join([serialize_fx_arg(arg) for arg in fx_args])
+    hashed_fx_args = base64.b32encode(hashlib.sha256(serialized_fx_args).digest())[:51]
+    return "f" + hashed_fx_args.decode("utf-8").lower()
+
+
+def get_path(basename: str, extension: str):
     subdir = os.path.join(cache_dir(), basename[1:3])
-    path = os.path.join(subdir, f"{basename}.{ext}")
-    return extra + basename, subdir, path
+    path = os.path.join(subdir, f"{basename}.{extension}")
+    return basename, subdir, path
 
 
-def write(source_code: Union[str, bytes], ext: str, extra="", binary=False):
-    basename, subdir, path = get_code_path(source_code, ext, extra, binary)
+def write(content: Union[str, bytes], key: str, extension: str):
+    basename, subdir, path = get_path(key, extension)
     if not os.path.exists(subdir):
         os.makedirs(subdir, exist_ok=True)
     if not os.path.exists(path):
-        write_atomic(path, source_code, binary)
+        write_atomic(path, content)
     return basename, path
 
 
-def write_atomic(path: str, source_code: Union[str, bytes], binary=False):
+def write_atomic(path: str, content: Union[str, bytes]):
     # Write into temporary file first to avoid conflicts between threads
     # Avoid using a named temporary file, as those have restricted permissions
+    assert isinstance(
+        content, (str, bytes)
+    ), "Only strings and byte arrays can be saved in the cache"
     path = pathlib.Path(path)
     tmp_path = path.parent / f".{os.getpid()}.{threading.get_ident()}.tmp"
-    with tmp_path.open("wb" if binary else "w") as f:
-        f.write(source_code)
-
+    write_mode = "w" if isinstance(content, str) else "wb"
+    with tmp_path.open(write_mode) as f:
+        f.write(content)
     tmp_path.rename(path)
 
 
+class FxGraphCache:
+    cache = dict()
+    clear = staticmethod(cache.clear)
+
+    @classmethod
+    def load(
+        cls, compile_fx_fn: Callable, fx_args: List[Any], fx_kwargs: Dict[str, Any]
+    ):
+        fx_args_for_hashing = copy(fx_args)
+        fx_args_for_hashing.extend(list(fx_kwargs.values()))
+        # Hash also on torch version and current triton config
+        fx_args_for_hashing.append(torch.__version__)
```

Comment:
can we use `torch.__version__` for cache? at least for local developments I would expect that version to be the same but I would also expect the cache to be invalidated

Reply:
for this exact case, @ezyang said that the cache should be off by default so that people developing locally don't hit the issue of cache invalidation. Another solution he proposed is making the cache have an index that can be set and serves to invalidate it all if you want to. Any thoughts on this?